### ISSUE TITLE
makeup transfer doesn't work

### ISSUE THREAD

**Issue by Dratlan** [basiclab/MAD#2]
Hi great work, I have some trouble in test makeup transfer B.3, I've download all the ckpts and organize my data as B.3 says, and it did run successfully. But the generated image is nearly the original source input. I want to figure out something happened, will u
 help me>=<?

**Comment by Dratlan** [comment]
Oh by the way, If I need retraining and refine this work to get a  512x512 output?

**Comment by Justin900429** [comment]
Hi, Thanks for the issue, could you please try again? I have updated the usage or you can check below. You should fine-tune the model to get the 512x512 output or at least use some scaling trick with modified timestep to scale up the image.

```shell
python generate_transfer.py \
    --config configs/model_256_256.yaml \
    --save-folder transfer_result \
    --source-root data/mtdataset/images \
    --target-root data/beautyface/images \
    --source-list assets/nomakeup.txt \
    --target-list assets/beauty_makeup.txt \
    --source-label 1 \
    --target-label 0 \
    --num-process 1 \
    --inpainting \
    --cam \
    --opts MODEL.PRETRAINED Justin900/MAD
```

**Comment by Dratlan** [comment]
@Justin900429 My usage is:
python generate_transfer.py --config configs/model_256_256.yaml --save-folder makeup_result --source-root makeup/id --target-root makeup/ref --source-list makeup/id1.txt --target-list makeup/ref1.txt --source-label 1 --target-label 0  --inpainting  --num-process 1  --cam --opts MODEL.PRETRAINED /workspace/nas-data/pretrained/MAD.
It run successfully, but the generate images change nothing.
May be I should run the example for checking.

**Comment by Justin900429** [comment]
If the makeup results are not obvious, maybe try changing the weights or the steps.
For example, increasing the weight and reducing the step can strengthen the makeup.

[Edit]
Another issue can be the wrong values of the mask. The value of the mask should correspond to the one in MT Dataset. Please check below:

| Value | Component         |
|-------|------------------|
| 9     | up lip           |
| 4     | face             |
| 8     | nose             |
| 10    | neck             |
| 12    | hair             |
| 1     | right eye        |
| 6     | left eye         |
| 2     | right eyebrow    |
| 7     | left eyebrow     |
| 5     | right ear        |
| 3     | left ear         |
| 0     | ear ring         |
| 11    | teeth            |
| 0     | shirt            |

**Comment by Dratlan** [comment]
I'll try for increasing the weights, where can i increasing the weight and reducing the step?
And about the mask, It seems I didn't need to offer the mask, so how can i check the value of the mask?

**Comment by Dratlan** [comment]
![Image](https://github.com/user-attachments/assets/57f27209-c195-4a54-8bd3-1f02ebce457d)
I think it's in here translation.py

**Comment by Justin900429** [comment]
Yeah, you can modify the scale and step with those variables. The scale is recommended to be within `[0, 1]` (not sure the outcome larger than 1).

The mask is automatically fetched. The path should be:

```
{img_root}
| - image
| - parsing
```

Once an image is given, it will search for the mask under the `parsing` folder.
It seems this is not that convenient for the customized image. Maybe I'll make it more flexible at some point.

**Comment by Dratlan** [comment]
I think maybe the mask is the matter(the scale changes nothing), I didn't find the parsing folder. The same not-changing results with text_editing results.

**Comment by Justin900429** [comment]
A correct mask is a must for performing transfer since we check the value to figure out where transfer should take place.  Maybe you can try creating a folder structure as shown above and provide the mask with the correct value.

**Comment by Dratlan** [comment]
Ok, so the mask is automatically fetched failed, I need to prepare the mask image for my source and ref image and organize it. And the value of the mask is the pixel value i think, am i right? is there sth ref or i can just do it myself.

**Comment by Justin900429** [comment]
Maybe try [face parsing](https://huggingface.co/jonathandinu/face-parsing) and convert the id to the correct one. It is recommended that the mask be saved in PNG format.

**Comment by Dratlan** [comment]
got it!

**Comment by eb-codes** [comment]
Did you find a way to make it work?

**Comment by Justin900429** [comment]
You can try the following script to obtain the mask for a customized image. I didn't test it but it should work though.

```python
import numpy as np
import torch
from PIL import Image
from torch import nn
from transformers import SegformerForSemanticSegmentation, SegformerImageProcessor

CONVERT_DICT = {
    11: 9,  # up lip
    1: 4,  # face
    2: 8,  # nose
    17: 10,  # neck
    16: 10,  # neck
    13: 12,  # hair
    5: 1,  # right eye
    4: 6,  # left eye
    7: 2,  # right eyebrow
    6: 7,  # left eyebrow
    9: 5,  # right ear
    8: 3,  # left ear
    15: 0,  # ear ring
    10: 11,  # teeth
    18: 0,  # shirt
}


def main(img_path: str, save_path: str):
    device = "cuda" if torch.cuda.is_available() else "cpu"
    image_processor = SegformerImageProcessor.from_pretrained(
        "jonathandinu/face-parsing"
    )
    model = SegformerForSemanticSegmentation.from_pretrained(
        "jonathandinu/face-parsing"
    )
    model.to(device)

    image = Image.open(img_path).convert("RGB")
    inputs = image_processor(images=image, return_tensors="pt").to(device)
    outputs = model(**inputs)
    logits = outputs.logits
    upsampled_logits = nn.functional.interpolate(
        logits,
        size=image.size[::-1],
        mode="bilinear",
        align_corners=False,
    )
    labels = upsampled_logits.argmax(dim=1)[0]
    labels = labels.cpu().numpy()
    new_labels = np.copy(labels)
    for key, value in CONVERT_DICT.items():
        new_labels[labels == key] = value
    new_labels = Image.fromarray(new_labels)
    new_labels.save(save_path)


if __name__ == "__main__":
    main("xxxx", "xxxx.png")
```

**Comment by Dratlan** [comment]
> Did you find a way to make it work?

You need to prepare the parsing map(mask) for your source and id image, and organize them.

**Comment by Dratlan** [comment]
> You can try the following script to obtain the mask for a customized image. I didn't test it but it should work though.
> 
> import numpy as np
> import torch
> from PIL import Image
> from torch import nn
> from transformers import SegformerForSemanticSegmentation, SegformerImageProcessor
> 
> CONVERT_DICT = {
>     11: 9,  # up lip
>     1: 4,  # face
>     2: 8,  # nose
>     17: 10,  # neck
>     16: 10,  # neck
>     13: 12,  # hair
>     5: 1,  # right eye
>     4: 6,  # left eye
>     7: 2,  # right eyebrow
>     6: 7,  # left eyebrow
>     9: 5,  # right ear
>     8: 3,  # left ear
>     15: 0,  # ear ring
>     10: 11,  # teeth
>     18: 0,  # shirt
> }
> 
> 
> def main(img_path: str, save_path: str):
>     device = "cuda" if torch.cuda.is_available() else "cpu"
>     image_processor = SegformerImageProcessor.from_pretrained(
>         "jonathandinu/face-parsing"
>     )
>     model = SegformerForSemanticSegmentation.from_pretrained(
>         "jonathandinu/face-parsing"
>     )
>     model.to(device)
> 
>     image = Image.open(img_path).convert("RGB")
>     inputs = image_processor(images=image, return_tensors="pt").to(device)
>     outputs = model(**inputs)
>     logits = outputs.logits
>     upsampled_logits = nn.functional.interpolate(
>         logits,
>         size=image.size[::-1],
>         mode="bilinear",
>         align_corners=False,
>     )
>     labels = upsampled_logits.argmax(dim=1)[0]
>     labels = labels.cpu().numpy()
>     new_labels = np.copy(labels)
>     for key, value in CONVERT_DICT.items():
>         new_labels[labels == key] = value
>     new_labels = Image.fromarray(new_labels)
>     new_labels.save(save_path)
> 
> 
> if __name__ == "__main__":
>     main("xxxx", "xxxx.png")

the input source image need to customized? I've prepare the parsing map, and there are some makeup in the result. But failed makeup, I'm trying to figure out this. the result of below is (source,ref,output):

![Image](https://github.com/user-attachments/assets/feeedbba-903a-48be-9b26-f72d5529b9ad)

![Image](https://github.com/user-attachments/assets/78ba98df-06a4-43bf-a359-fe893ccf3102)

![Image](https://github.com/user-attachments/assets/c7bb6196-0c13-4aee-8739-2da4997c61cf)

**Comment by Justin900429** [comment]
```python
import argparse

import numpy as np
import torch
from PIL import Image
from torch import nn
from transformers import SegformerForSemanticSegmentation, SegformerImageProcessor

CONVERT_DICT = {
    12: 13,  # down lip
    11: 9,  # up lip
    1: 4,  # face
    2: 8,  # nose
    17: 10,  # neck
    16: 10,  # neck
    13: 12,  # hair
    5: 1,  # right eye
    4: 6,  # left eye
    7: 2,  # right eyebrow
    6: 7,  # left eyebrow
    9: 5,  # right ear
    8: 3,  # left ear
    15: 0,  # ear ring
    10: 11,  # teeth
    18: 0,  # shirt
}


def main(img_path: str, save_path: str):
    device = "cuda" if torch.cuda.is_available() else "cpu"
    image_processor = SegformerImageProcessor.from_pretrained(
        "jonathandinu/face-parsing"
    )
    model = SegformerForSemanticSegmentation.from_pretrained(
        "jonathandinu/face-parsing"
    )
    model.to(device)

    image = Image.open(img_path).convert("RGB")
    inputs = image_processor(images=image, return_tensors="pt").to(device)
    outputs = model(**inputs)
    logits = outputs.logits
    upsampled_logits = nn.functional.interpolate(
        logits,
        size=image.size[::-1],
        mode="bilinear",
        align_corners=False,
    )
    labels = upsampled_logits.argmax(dim=1)[0]
    labels = labels.cpu().numpy()
    new_labels = np.copy(labels)
    for key, value in CONVERT_DICT.items():
        new_labels[labels == key] = value
    new_labels = new_labels.astype("uint8")
    new_labels = Image.fromarray(new_labels)
    new_labels.save(save_path)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--img_path",
        type=str,
        required=True,
        help="Path to the input image.",
    )
    parser.add_argument(
        "--save_path",
        type=str,
        required=True,
        help="Path to save the output image.",
    )
    args = parser.parse_args()
    main(args.img_path, args.save_path)
```

We have used the provided image to get the parsing mask from both the source and reference images. The above script should work correctly. Here's how we organize the image.

## 1. Create the parsing mask

```
{root}
| - images
| - | - makeup/test.png (whatever name)
| - | - nonmakeup
| - parsing
| - | - makeup/test.png (whatever name)
| - | - nonmakeup
```

(You can create the parsing mask with the following commands)
```shell
python face_parsing_script.py --img_path {root}/makeup/images/test.png --save_path {root}/makeup/parsing/test.png 
python face_parsing_script.py --img_path {root}/makeup/images/test.png --save_path {root}/makeup/parsing/test.png 
```

Here is the finalized folder structure

```
{root}
| - images
| - | - makeup/test.png
| - | - nonmakeup/test.png
| - parsing
| - | - makeup/test.png
| - | - nonmakeup/test.png
```

## 2. Create the txt file

Afterward, create two `txt` files called `makeup.txt` and `nonmakeup.txt` with the following content:

```plain
# makeup.txt
makeup/test.png
# nonmakeup.txt
nonmakeup/test.png
```

## 3. Run the makeup transfer

We use the following command:

```shell
python generate_transfer.py \  
    --config configs/model_256_256.yaml \
    --save-folder temp_test \
    --source-root {root}/images \
    --target-root {root}/images \
    --source-list nomakeup.txt \
    --target-list makeup.txt \
    --source-label 1 \
    --target-label 0 \
    --num-process 1 \
    --inpainting \
    --cam \
    --opts MODEL.PRETRAINED Justin900/MAD
```

Here are the outcomes:

![Image](https://github.com/user-attachments/assets/cef3d7fa-1ce5-4a56-9962-bc9ca55dfde3)

> Some analysis of the results:
> As we pointed out in our paper, our model only learns to generate makeup and non-makeup images on about a thousand images. Since we only trained with them, our model might **overfit** those images as we trained for several iterations and might only remember what those images are. This might explain why the customized image might not work (not the style; the unseen style can work). A promising way to make it work might be training on more human faces to make the learned distribution diverse and big enough to cover different faces.

**Comment by Dratlan** [comment]
Thanks for your detail demostrartion
